### DISCUSSION TITLE
Acquire the Refresh Token on Ubuntu Server

### DISCUSSION THREAD

**Discussion by morrisjosephd** [tsightler/ring-mqtt#837] [Support Q&A] [answered]
I'm attempting to setup `ring-mqtt` as a docker container on a machine running `Ubuntu 22.04.4 LTS`.  I have followed the installation instructions on the WIKI.

However running `docker run -it --rm --mount type=bind,source=<host_path>,target=/data --entrypoint /app/ring-mqtt/init-ring-mqtt.js tsightler/ring-mqtt` is hanging on my server.

I know `node` is running because I am asked for my Email and Password which I provide.  Then the script hangs.  When I run `docker logs <container>` the only logs I get back is the Email and Password I have provided.

Of interest, when I execute the same WIKI instructions using Docker on my Mac, I get the expected results of being asked for my `2FA Code` and `ring-state.json` is created in the `/data` directory.

I'm seeking assistance on getting past the Refresh Token (init-ring-mqtt.js) script hanging on my Ubuntu machine.

**Comment by tsightler** [comment]
I mean, if you are successfully able to create a config from the other machine why not just copy it over?  To me that would be the easiest way.  No telling what it might be otherwise.

**Reply by morrisjosephd** [reply]
That was a good idea, but unfortunately it didn't work. I'm able to get the `ring-mqtt`  container to start as _now_ I have the requisite files in the `/data` directory (`config.json` and `ring-state.json`) copied from my mac.  However, the log files indicate a hang at the Ring API:

```bash
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
cont-init: info: running /etc/cont-init.d/ring-mqtt.sh
-------------------------------------------------------
| Ring-MQTT with Video Streaming                      |
| Docker Edition                                      |
|                                                     |
| For support questions please visit:                 |
| https://github.com/tsightler/ring-mqtt/discussions  |
-------------------------------------------------------
cont-init: info: /etc/cont-init.d/ring-mqtt.sh exited 0
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
services-up: info: copying legacy longrun ring-mqtt (no readiness notification)
s6-rc: info: service legacy-services successfully started
-------------------------------------------------------
ring-mqtt.js version: 5.6.4
Node version v18.20.1
NPM version 9.6.6
git version 2.40.1
-------------------------------------------------------
Running ring-mqtt...
2024-05-22T18:06:47.086Z ring-mqtt Detected runmode: docker
2024-05-22T18:06:47.088Z ring-mqtt Configuration file: /data/config.json
2024-05-22T18:06:48.060Z ring-mqtt Reading latest data from state file: /data/ring-state.json
2024-05-22T18:06:48.064Z ring-mqtt MQTT URL: mqtt://localhost:1883
2024-05-22T18:06:48.135Z ring-mqtt Attempting connection to Ring API using saved refresh token...
```

Things I know:
- The request is making it's way out of the container as I see the request in my DNS query logs.
-  Along those lines, when I enter the container I am able to ping a DNS resolver from it.
- With the container running, it makes requests DNS requests (a lot of them) to `oauth.ring.com`. 

I have run out of troubleshooting ideas.  Do you see anything in the logs or have any other ideas?

**Reply by tsightler** [reply]
My guess is you have a VPN, or some other security setup, that Ring is blocking, or that is blocking the request, but it's just a guess.  It's simple network stuff here, nothing more, I can't really help with that.

**Reply by morrisjosephd** [reply, marked as answer]
Yep, it was my own network.  I'm running `Pi-hole` in a container on the same box that I'm running `ring-mqtt` in a container.  My thinking is though the network-as-a-whole DNS was working properly, the container-to-container DNS was NOT working properly.  

This [forum post](https://discourse.pi-hole.net/t/solved-dns-resolution-is-currently-unavailable/33725/11) was the solution for me.

**Reply by tsightler** [reply]
A DNS Haiku:

> It’s not DNS
> There’s no way it’s DNS
> It was DNS

It's always DNS, but f it's not DNS, it's probably the network and, if it's not the network, it's probably an expired certificate, but if it's not an expired certificate, well, it's probably DNS.
